### PR TITLE
DetailedCardView

### PR DESCRIPTION
**Pull request by kuttnig**

### CODE REVIEW

**Comment by raphaelseher** [comment]
|File|Coverage [0%]|:green_apple:|
|:-|:-:|:-:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/1d8fc33e909ac7e9c00449f400379331e7d95e43/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/1d8fc33e909ac7e9c00449f400379331e7d95e43/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|

|Total Project Coverage|0%|:green_apple:|
|:-|:-:|:-:|

**Comment by raphaelseher** [comment]
|File|Coverage [0%]|:green_apple:|
|:-|:-:|:-:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/877b92a0103dc74c3d58b2cc322e628438061e23/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/877b92a0103dc74c3d58b2cc322e628438061e23/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|
|[MainActivity.java](https://github.com/SE2-Gwent/Gwent/blob/877b92a0103dc74c3d58b2cc322e628438061e23/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2FMainActivity.java)|0%|:green_apple:|

|Total Project Coverage|0%|:green_apple:|
|:-|:-:|:-:|

**Comment by raphaelseher** [comment]
|File|Coverage [0%]|:green_apple:|
|:-|:-:|:-:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/7d183abbb171d69068fe7a05a096d9db8a10e555/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/7d183abbb171d69068fe7a05a096d9db8a10e555/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|
|[Card.java](https://github.com/SE2-Gwent/Gwent/blob/7d183abbb171d69068fe7a05a096d9db8a10e555/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fdata%2FCard.java)|0%|:green_apple:|
|[MainActivity.java](https://github.com/SE2-Gwent/Gwent/blob/7d183abbb171d69068fe7a05a096d9db8a10e555/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2FMainActivity.java)|0%|:green_apple:|

|Total Project Coverage|0%|:green_apple:|
|:-|:-:|:-:|

**Comment by raphaelseher** [comment]
|File|Coverage [0%]|:green_apple:|
|:-|:-:|:-:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/bea7eeb53e6c531e321672b1b86b794eb6dca5e1/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/bea7eeb53e6c531e321672b1b86b794eb6dca5e1/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|
|[Card.java](https://github.com/SE2-Gwent/Gwent/blob/bea7eeb53e6c531e321672b1b86b794eb6dca5e1/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fdata%2FCard.java)|0%|:green_apple:|
|[MainActivity.java](https://github.com/SE2-Gwent/Gwent/blob/bea7eeb53e6c531e321672b1b86b794eb6dca5e1/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2FMainActivity.java)|0%|:green_apple:|

|Total Project Coverage|0%|:green_apple:|
|:-|:-:|:-:|

**Comment by raphaelseher** [comment]
|File|Coverage [0%]|:green_apple:|
|:-|:-:|:-:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/e338c0c483ce548c60cc5008f10c70f804d635f4/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/e338c0c483ce548c60cc5008f10c70f804d635f4/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|
|[Card.java](https://github.com/SE2-Gwent/Gwent/blob/e338c0c483ce548c60cc5008f10c70f804d635f4/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fdata%2FCard.java)|0%|:green_apple:|
|[MainActivity.java](https://github.com/SE2-Gwent/Gwent/blob/e338c0c483ce548c60cc5008f10c70f804d635f4/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2FMainActivity.java)|0%|:green_apple:|

|Total Project Coverage|0%|:green_apple:|
|:-|:-:|:-:|

**Comment by raphaelseher** [comment]
|File|Coverage [0%]|:green_apple:|
|:-|:-:|:-:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/c79a91f09dad1ba6aaf0971fc97780beab55f2eb/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/c79a91f09dad1ba6aaf0971fc97780beab55f2eb/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|
|[Card.java](https://github.com/SE2-Gwent/Gwent/blob/c79a91f09dad1ba6aaf0971fc97780beab55f2eb/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fdata%2FCard.java)|0%|:green_apple:|
|[MainActivity.java](https://github.com/SE2-Gwent/Gwent/blob/c79a91f09dad1ba6aaf0971fc97780beab55f2eb/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2FMainActivity.java)|0%|:green_apple:|

|Total Project Coverage|0%|:green_apple:|
|:-|:-:|:-:|

**Comment by raphaelseher** [comment]
|File|Coverage [0%]|:green_apple:|
|:-|:-:|:-:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/8a2ec83f52af10b11d69dfa2bdf7aa29632fb37e/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/8a2ec83f52af10b11d69dfa2bdf7aa29632fb37e/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|
|[Card.java](https://github.com/SE2-Gwent/Gwent/blob/8a2ec83f52af10b11d69dfa2bdf7aa29632fb37e/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fdata%2FCard.java)|0%|:green_apple:|
|[MainActivity.java](https://github.com/SE2-Gwent/Gwent/blob/8a2ec83f52af10b11d69dfa2bdf7aa29632fb37e/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2FMainActivity.java)|0%|:green_apple:|

|Total Project Coverage|0%|:green_apple:|
|:-|:-:|:-:|

**Comment by raphaelseher** [comment]
|File|Coverage [0%]|:green_apple:|
|:-|:-:|:-:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/c68e8c5a0e834e3975e85be83935e7a3a8e95b45/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/c68e8c5a0e834e3975e85be83935e7a3a8e95b45/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|
|[MainActivity.java](https://github.com/SE2-Gwent/Gwent/blob/c68e8c5a0e834e3975e85be83935e7a3a8e95b45/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2FMainActivity.java)|0%|:green_apple:|

|Total Project Coverage|0%|:green_apple:|
|:-|:-:|:-:|

**Comment by kuttnig** [comment]
@mavalo99, @raphaelseher 
Soweit wäre das Fragment für DetailedCardView fertig.
Die Idee besteht darin, dass mit einem Klick auf den LongClickListener, welchen @mavalo99 implementiert hat, die ID der Karte übergeben wird. Das ViewModel holt sich dann über die ID alle Daten, welche das Fragment benötigt (CardText, CardName, ...) aus dem zugehörigen JSON-file. Um das Img anzuzeigen, befindet sich im JSON-file der Name der Resource, welche benötigt wird. Bitte gerne Verbesserungsvorschläge/Änderungen einbringen. Lg

**Comment by raphaelseher** [comment]
|File|Coverage [5.23%]|:green_apple:|
|:-|:-:|:-:|
|[Card.java](https://github.com/SE2-Gwent/Gwent/blob/f9bf1e9372f55cac0d3e6884a41ae4e729a47e37/gamelogic%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgamelogic%2Fmodels%2FCard.java)|37.65%|:green_apple:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/f9bf1e9372f55cac0d3e6884a41ae4e729a47e37/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/f9bf1e9372f55cac0d3e6884a41ae4e729a47e37/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|
|[MainActivity.java](https://github.com/SE2-Gwent/Gwent/blob/f9bf1e9372f55cac0d3e6884a41ae4e729a47e37/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2FMainActivity.java)|0%|:green_apple:|

|Total Project Coverage|23.81%|:green_apple:|
|:-|:-:|:-:|

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -13,7 +13,8 @@
             android:name=".MainActivity"
             android:exported="true"
             android:label="@string/app_name"
-            android:theme="@style/Theme.Gwent.NoActionBar">
+            android:theme="@style/Theme.Gwent.NoActionBar"
+            android:screenOrientation="landscape">
```
Comment (raphaelseher) [inline app/src/main/AndroidManifest.xml:17]: Dann hamma natürlich den StartScreen auch in Landscape, wäre besser, wenn wir zwischen Portrait/Landscape wechseln im Code.


https://stackoverflow.com/questions/37034436/change-android-screen-orientation-from-code

File: app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardFragment.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    binding = DetailedCardFragmentBinding.inflate(inflater, container, false);
+    return binding.getRoot();
+  }
+
+  @Override
+  public void onActivityCreated(@Nullable Bundle savedInstanceState) {
+    super.onActivityCreated(savedInstanceState);
+
+    // this should destroy the fragment if the button is clicked
+    binding.backgroundImageButton.setOnClickListener(
```
Comment (raphaelseher) [inline app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardFragment.java:55]: FYI: du könntest auch eine ImageView mit dem Attribute `clickable` verwenden, dann würds auch gehen.

Comment (raphaelseher) [inline app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardFragment.java:55]: FYI: du könntest auch eine ImageView mit dem Attribute `clickable` verwenden, dann würds auch gehen.

File: app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardFragment.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        view -> {
+          FragmentManager fragmentManager = getActivity().getSupportFragmentManager();
```
Comment (raphaelseher) [inline app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardFragment.java:57]: FYI: Mit 
``` 
Navigation.findNavController(
                  Objects.requireNonNull(getActivity()), R.id.nav_host_fragment_content_main)
              .popBackStack();
```
würden wir wieder retour zum StartFragment kommen. 
In der GameBoard View sollten wir dann aber die Fragments übereinander anzeigen, also nicht wichtig.

File: app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardFragment.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        binding.cardPower.setText(String.valueOf(viewModel.getCard().getPower()));
+
+        // we have to convert the enum-array to a String, before setting the textView
+        setTextViewCardTypes(viewModel.getCard().getTypes());
+    }
+  }
+
+  // update textColor of textView power according to powerDiff
+  private void updateColorCardPower(int powerDiff) {
+    if (powerDiff == 0) {
```
Comment (raphaelseher) [inline app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardFragment.java:82]: Spar ma uns das if: 
```
 @ColorRes int color = (powerDiff == 0) ? R.color.green : R.color.red;
    binding.cardPower.setTextColor(getResources().getColor(color, null));
```

File: app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardViewModel.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import at.aau.se2.gamelogic.models.CardType;
+
+public class DetailedCardViewModel extends ViewModel {
+  private Card card;
+  private String imgResourceName;
+
+  private MutableLiveData<DetailedCardViewModel.ViewState> currentState =
+      new MutableLiveData<>(DetailedCardViewModel.ViewState.INITIAL);
+
+  enum ViewState {
```
Comment (raphaelseher) [inline app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardViewModel.java:18]: Wir werden da keine Loading States brauchen, aber wir können das ViewModel besser verwenden.
Mach aus dem ViewState eine Klasse mit den fertigen Properties, die wir in der View dann brauchen.
Also z.B. mit `String cardViewTypes` und bringen dann die Logic dafür auch ins ViewModel.

File: app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardFragment.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+      binding.cardPower.setTextColor(getResources().getColor(R.color.green, null));
+    } else {
+      binding.cardPower.setTextColor(getResources().getColor(R.color.red, null));
+    }
+  }
+
+  // set textView cardTypes (convert Array of enums to string)
+  private void setTextViewCardTypes(ArrayList<CardType> types) {
```
Comment (raphaelseher) [inline app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardFragment.java:90]: Ins ViewModel, siehe nächster Kommentar.

File: app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardViewModel.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  we reference resources according to detailed-card-view by taking the name of the card
+  first we replace all spaces by an underscore
+  second we convert the string to lowercase
+  third we append '_detail to it'
+
+  for example:
+  Ard Feainn Crossbow Man -> ard_feainn_crossbow_man_detail
+   */
+  private void setImgResourceName() {
+    String imageResourceName = "";
```
Comment (raphaelseher) [inline app/src/main/java/at/aau/se2/gwent/views/detailedcard/DetailedCardViewModel.java:71]: Die Bilder sollten wir bei den Cards in die Properties verlinken. Vl. checkst das beim Static Data Task dann mit.

**Comment by kuttnig** [comment]
@raphaelseher thx ich werde heute/morgen mal drüber schauen und ggf. Änderungen vornehmen. Ich ping dich dann nochmal. Lg

**Comment by raphaelseher** [comment]
|File|Coverage [7.82%]|:green_apple:|
|:-|:-:|:-:|
|[GameField.java](https://github.com/SE2-Gwent/Gwent/blob/996268e405077e01bd86dd87e31e9248ba6f6d1c/gamelogic%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgamelogic%2Fmodels%2FGameField.java)|65.85%|:green_apple:|
|[Card.java](https://github.com/SE2-Gwent/Gwent/blob/996268e405077e01bd86dd87e31e9248ba6f6d1c/gamelogic%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgamelogic%2Fmodels%2FCard.java)|37.65%|:green_apple:|
|[CardDetails.java](https://github.com/SE2-Gwent/Gwent/blob/996268e405077e01bd86dd87e31e9248ba6f6d1c/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FCardDetails.java)|0%|:green_apple:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/996268e405077e01bd86dd87e31e9248ba6f6d1c/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/996268e405077e01bd86dd87e31e9248ba6f6d1c/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|
|[Hero.java](https://github.com/SE2-Gwent/Gwent/blob/996268e405077e01bd86dd87e31e9248ba6f6d1c/gamelogic%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgamelogic%2Fmodels%2FHero.java)|0%|:green_apple:|
|[MainActivity.java](https://github.com/SE2-Gwent/Gwent/blob/996268e405077e01bd86dd87e31e9248ba6f6d1c/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2FMainActivity.java)|0%|:green_apple:|

|Total Project Coverage|23.18%|:green_apple:|
|:-|:-:|:-:|

**Review by mavalo99 — APPROVED** [review]
-> läuft
-> Back-Button Issue

**Review by raphaelseher — APPROVED** [review]

**Comment by raphaelseher** [comment]
|File|Coverage [7.69%]|:green_apple:|
|:-|:-:|:-:|
|[GameField.java](https://github.com/SE2-Gwent/Gwent/blob/d09f89475a72af6f658a8fb55a27757e2226c915/gamelogic%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgamelogic%2Fmodels%2FGameField.java)|65.85%|:green_apple:|
|[Card.java](https://github.com/SE2-Gwent/Gwent/blob/d09f89475a72af6f658a8fb55a27757e2226c915/gamelogic%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgamelogic%2Fmodels%2FCard.java)|37.65%|:green_apple:|
|[CardDetails.java](https://github.com/SE2-Gwent/Gwent/blob/d09f89475a72af6f658a8fb55a27757e2226c915/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FCardDetails.java)|0%|:green_apple:|
|[DetailedCardFragment.java](https://github.com/SE2-Gwent/Gwent/blob/d09f89475a72af6f658a8fb55a27757e2226c915/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardFragment.java)|0%|:green_apple:|
|[DetailedCardViewModel.java](https://github.com/SE2-Gwent/Gwent/blob/d09f89475a72af6f658a8fb55a27757e2226c915/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2Fviews%2Fdetailedcard%2FDetailedCardViewModel.java)|0%|:green_apple:|
|[Hero.java](https://github.com/SE2-Gwent/Gwent/blob/d09f89475a72af6f658a8fb55a27757e2226c915/gamelogic%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgamelogic%2Fmodels%2FHero.java)|0%|:green_apple:|
|[MainActivity.java](https://github.com/SE2-Gwent/Gwent/blob/d09f89475a72af6f658a8fb55a27757e2226c915/app%2Fsrc%2Fmain%2Fjava%2Fat%2Faau%2Fse2%2Fgwent%2FMainActivity.java)|0%|:green_apple:|

|Total Project Coverage|23.04%|:green_apple:|
|:-|:-:|:-:|

**Pull request merged**